tests/unit_tests/data/test_preprocess_data.py

```diff
@@ -184,25 +209,70 @@ def gpt2_merge(odir):
     return path
 
 
+def gpt_args(temp_dir):
+    # gpt specific args
+    return [
+        "--tokenizer-type",
+        "GPT2BPETokenizer",
+        "--vocab-file",
+        gpt2_vocab(temp_dir),
+        "--merge-file",
+        gpt2_merge(temp_dir),
+        "--append-eod",
+        "--workers",
+        "10",
+        "--log-interval",
+        "1",
+    ]
+
+
 def test_preprocess_data_gpt():
     with tempfile.TemporaryDirectory() as temp_dir:
+        do_test_preprocess_data(temp_dir, extra_args=gpt_args(temp_dir))
 
-        # gpt specific args
-        gpt_args = [
-            "--tokenizer-type",
-            "GPT2BPETokenizer",
-            "--vocab-file",
-            gpt2_vocab(temp_dir),
-            "--merge-file",
-            gpt2_merge(temp_dir),
-            "--append-eod",
-            "--workers",
-            "10",
-            "--log-interval",
-            "1",
-        ]
 
-        do_test_preprocess_data(temp_dir, extra_args=gpt_args)
+class MockS3Client:
+    def __init__(self, *args, **kwargs):
+        self._data = {}
+
+    def download_file(self, bucket, key, local_path):
+        with open(local_path, "wb") as fout:
+            fout.write(self._data[(bucket, key)])
+
+    def upload_file(self, local_path, bucket, key):
+        with open(local_path, "rb") as fin:
+            self._data[(bucket, key)] = fin.read()
+
+    def head_object(self, Bucket, Key):
+        return (Bucket, Key) in self._data
+
+    def get_object(self, Bucket, Key, Range):
+        assert Range.startswith("bytes=")
+        parts = Range.split("=")
+        assert len(parts) == 2
+        subparts = parts[1].split("-")
+        assert len(subparts) == 2
+        start = int(subparts[0])
+        # add 1 to convert inclusive index into exclusive index.
+        end = int(subparts[1]) + 1
+        return {"Body": io.BytesIO(self._data[(Bucket, Key)][start:end])}
+
+    def close(self):
+        pass
+
+
+def test_preprocess_data_gpt_s3(monkeypatch):
+    MOCK_S3_CLIENT = MockS3Client()
+
+    def mock_s3_client(*args, **kwargs):
+        return MOCK_S3_CLIENT
+
+    monkeypatch.setattr("boto3.client", mock_s3_client)
```

Comment:
Why would the following not work?

```
monkeypatch.setattr("boto3.client", MockS3Client)
```

Reply:
It doesn't work, because it creates a new MockS3Client each time boto3.client is called. Instead, I want the state (i.e., `self._data`) to be maintained across calls to boto3.client, because it reflects the state of the mock S3.